Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c

```diff
@@ -0,0 +1,705 @@
+/** @file
+*
+*  Copyright (c) 2019 Microsoft Corporation. All rights reserved.
+*
+*  This program and the accompanying materials
+*  are licensed and made available under the terms and conditions of the BSD License
+*  which accompanies this distribution.  The full text of the license may be found at
+*  http://opensource.org/licenses/bsd-license.php
+*
+*  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+*
+**/
+
+#include "SmbiosConfigDxe.h"
+
+#define MAX_SMBIOS_KEY_REV_1 5
+CHAR8* mSmbiosKeyRev1[MAX_SMBIOS_KEY_REV_1] = {"System Manufacturer",
+                                               "System Product Name",
+                                               "System SKU",
+                                               "System Family",
+                                               "Baseboard Product"};
+
+STATIC CONST CHAR16 mSmbiosOverridePresent[] = L"SmbiosOverridePresent";
+
+STATIC SMBIOS_OVERRIDE_NODE *mSmbiosOverrideListHead = NULL;
+
+/**
+
+  Finds and retrieves the next Key/Value pair from the buffer starting at the
+  given offset.
+
+  This routine scans through the provided buffer starting at the given offset.
+  It finds the first '"' character and records its location as the start of the Key.
+  Then it continues scanning to find the next '"' instance. The length denotes the KeyLength.
+
+  We repeat these steps to find the Value position and length.
+
+  At any point if a failure is detected, the algorithm bails with an error.
+
+  If we successfully find both Key and Value positions and lengths, then allocate/copy
+  these to their respective buffers and return to the caller.
+
+**/
+EFI_STATUS
+EFIAPI
+GetNextKeyValuePair (
+  CHAR8 *Buffer,
+  UINTN *Offset,
+  UINTN BufferSize,
+  CHAR8 **Key,
+  CHAR8 **Value
+  )
+{
+  EFI_STATUS Status;
+  UINTN CurrentOffset;
+  UINTN KeyLength;
+  UINTN ValueLength;
+  JSON_PARSER_STATE State;
+  CHAR8 *KeyStart;
+  CHAR8 *ValueStart;
+  CHAR8 *TempKey;
+  CHAR8 *TempValue;
+
+  CurrentOffset = *Offset;
```

Comment:
Check Offset, Buffer, and other parameters for NULL and return bad parameter error?